### ISSUE TITLE
Locally cached Context for OAv3 docs

### ISSUE THREAD

**Issue by morgatron** [Open-Attestation/open-attestation#234]
We're currently running OA as a serverless (Lambda) app, which breaks context pre-caching- every call to wrapDocument() fetches the 5 named jsonld contexts.

The obvious solution is to pull in these contexts at build time. Does this make sense? It's not obvious to me that dynamic loading is necessary at all, but I may missing something. If it is necessary, perhaps there could be a flag to check for changed contexts at run-time and load them (while perhaps giving a warning)?

I note that other VC libraries (e.g. DIDKit, Transmute... ) tend to have a stricter stance on what contexts are allowed. In DIDKit for instance they currently have to be included at (compile time)[https://spruceid.dev/docs/didkit/document-loading/], although pre-caching of a named subset may be allowed in the future. 

If this make sense, I would be happy to raise in a PR on this.

**Comment by morgatron** [comment]
(On a side note): 
This is related to two other points of interest which I'll flag here (I can open new issues if there's any interest):
1. I see that other's e.g. Transmute in their [examples](https://github.com/transmute-industries/verifiable-data/tree/main/packages/jsonld-document-loader) use their documentLoader for loading DIDs with DIF's [universal resolver](https://github.com/decentralized-identity/universal-resolver). This seems like an approach with some merit.
2. There doesn't seem to be any schema validation or JSON-LD validity checking on the verification end- documents are instead just checked against a TypeScript type. This surprises me a but- but perhaps this is deliberate...

**Comment by morgatron** [comment]
> Hi @morgatron, thanks for the detailed writeup! I will try to get the ball rolling to best of my ability. The rest of the OpenAttestation team may also chime in where necessary.

(Apologies if it was a bit of a dump!)


> Do you have any examples of such other software? I recall hearing you mention about a Transmute verifier over our earlier conversation. Perhaps you could link us to their verifier?

Some examples of other software (google search name + github will get you there for each of these):
* Transmute
* DIDkit
* Digital Bazaar
* DanubeTech

Of these, DIDkit's is probably the easiest to use, as they have a nice command line utility.

For installation-free testing, you can try:
* DanubeTech's ["universal verifier"](https://univerifier.io/)
* Transmute's https://api.did.actor

> Is this test suite similar to one laid out [here](https://w3c.github.io/vc-test-suite/implementations/)? There is a column for OpenAttestation describing the respective test results.

This is a subtle point. That test suite you link is purely for the VC data model. Applying it to OA simply checks that the correct fields exist in an issued doc and that no un-qualified data is in there. This is a great first step...

 **but**: it does _not_ check that the fields are used as the spec intended, OR that the methods used are understood by other software. 

The vc-api test suite I linked is a step toward interoperability testing. It assumes a given issuance/verification api (the 'vc-api'), and attempts to issue and verify generic VCs. We've already put a simple vc-api wrapper around OA and applied this test suite. I can share the initial results if you're itnerested.

In our meeting, Richard also mentioned a related test suite, fromthe w3c-ccg ['traceability interop' project](https://github.com/w3c-ccg/traceability-interop) and the SVIP 'plugfest'. This is an actual interop test between different VC issuers/verifiers, and is implemented as a set of postman collections calling the a vc-api. Ideally we should go green on this too.



> Do you mean [this code segment](https://github.com/Open-Attestation/oa-verify/blob/fd5d5f09e2817eb9e69a28c76654ee7a9bc5bae6/src/verifiers/documentStatus/didSigned/didSignedDocumentStatus.ts#L228)? I don't know the exact reason why the `issuer` field is left out during verification. Looking again at the [W3C specs](https://www.w3.org/TR/vc-data-model/#issuer), the `issuer.id` field can be a DID URL. Which means it should be set the same value as `openAttestationMetadata.proof.value`. Would you agree?

Yep, exactly! This is an example of where OA fits the VC data model (it has an 'issuer' field), but isn't practicaly interoperable (because it doesn't actually _use_ the 'issuer' field). The simple change of `openAtttestationMetadata.proof.value` -> `issuer.id` would do it from our point of view. But the caveat is I'm not sure how to support multiple issuers at this time, as you do with OAv2.

> Yes, it does sound like we should use `credentialSchema`. Need to confirm with the rest of the OA team if we can make the change.

Great!

> The OA verifier actually creates a [resolver](https://github.com/Open-Attestation/oa-verify/blob/fd5d5f09e2817eb9e69a28c76654ee7a9bc5bae6/src/did/resolver.ts#L32) that can resolve both `did:eth` and `did:web` URLs. However, the verifier can only [verify signatures signed using the `EcdsaSecp256k1RecoveryMethod2020`](https://github.com/Open-Attestation/oa-verify/blob/fd5d5f09e2817eb9e69a28c76654ee7a9bc5bae6/src/did/verifier.ts#L78-L113). That being said, there is also no toolkit offered by OA to manage the creation or hosting of `did:web` at this time.

Yes I'm sorry, I didn't do a good job in my description. I do know that OA uses the standard did:ethr resolver from DIF, and can use one for did:web too (which is great!). But as you say, it only supports signing a hash using EcdsaSecpk1. Which means that most did:web's used in the wild won't work , even when using OA's own proof method, because EcdsaSecpk1 is pretty rare outside of block-chain use cases. 

Have you put much thought in how best to do general DID resolution- e.g. using JS components compared to an external service implementing something like https://dev.uniresolver.io? An e.g of the latter is this [Transmute example](https://github.com/transmute-industries/verifiable-data/tree/main/packages/jsonld-document-loader). This would relate to #234.


> Yes, that is correct. Once `did:web` or `did:dns` methods are used, the `openAttestationMetadata.identityProof` becomes redundant as the act of resolving such DIDs will already prove (or map) the DID to a domain.

This is part of that verification discussion that I explained badly this morning. What does the verifier look like for such a document? I'm wondering if it's cleanest if a single verifier ccould return two fragments, so that verifying a did:dns signature could return DOCUMENT_INTEGRITY and ISSUER_IDENTITY... but perhaps I should write this on the other [issue](https://github.com/Open-Attestation/oa-verify/issues/230)!

**Comment by superical** [comment]
When we say the "context", can I presume it's referring to the URLs in the `@context` section of a document? If that's what it's referred to, would it make sense and probably easier for the apps to cache them locally on the users' clients instead of caching on the server side through a lambda? For eg, localstorage?

**Comment by morgatron** [comment]
Yep, I do mean the JSON-LD files referred to in the @context section.

>would it make sense and probably easier for the apps to cache them locally on the users' clients instead of caching on the server side through a lambda?

I realise now I haven't been clear above, sorry! 

We are wanting to use OpenAttestation behind the VC-API, largely because this API is becoming widely used. In this context, we want to run OA on the server. We're currently using AWS Lambda for this.

Does this make more sense now?
